lib/appInsightsWrapper.js

```diff
@@ -0,0 +1,171 @@
+  if (propertyDict.osDist && propertyDict.osRelease) {
+    return;
+  }
+  propertyDict.osDist = osDistro.dist;
+  propertyDict.osRelease = osDistro.release;
+}
+
+var dummyFunc = function () { };
+
+function initMacAndTrackEvent(eventName, properties) {
+  getmacs().then(function (allMacs) {
+    initMac(allMacs);
+  }).catch(dummyFunc).finally(function () {
+    internalTrackEvent(eventName, properties);
+  });
+}
+
+function initOsDistroAndTrackEvent(eventName, properties) {
+  getos().then(function (osDistro) {
+    initOsDistro(osDistro);
+  }).catch(dummyFunc).finally(function () {
+    internalTrackEvent(eventName, properties);
+  });
+}
+
+function initPropertiesAndTrackEvent(eventName, properties) {
+  getmacs().then(function (allMacs) {
+    initMac(allMacs);
+    return getos();
+  }).catch(function () {
+    return getos();
+  }).then(function (osDistro) {
+    initOsDistro(osDistro);
+  }).catch(dummyFunc).finally(function () {
+    internalTrackEvent(eventName, properties);
+  });
+}
+
+function internalTrackEvent(eventName, properties) {
+  // Add common properties
+  properties.parentModuleName = propertyDict.parentPkgName;
+  properties.parentModuleVersion = propertyDict.parentPkgVersion;
+  properties.packageVersion = propertyDict.currentPkgVersion;
+  properties.nodeVersion = process.version;
+  properties.osArch = os.arch();
+  properties.hostname = os.hostname();
+  properties.osType = os.type();
+  // Add subscription Id & tenant Id
+  properties.subscriptionId = propertyDict.subscriptionId;
+  properties.tenantId = propertyDict.tenantId;
+  // Add MAC
+  properties.mac = propertyDict.mac || [];
+  // Add OS Distro
+  properties.osPlatform = propertyDict.osDist || os.platform();
+  properties.osRelease = propertyDict.osRelease || os.release();
+  // Send event to application insights
+  appInsight.client.trackEvent(eventName, properties);
+}
+
+module.exports.start = function (instrumentationKey) {
+  // Skip asking user's approval for data collection for preview release
+  if (!settings.isBIEnabled(true)) {
+    return false;
+  }
+  appInsight.setup(instrumentationKey)
+    .setAutoCollectConsole(false)
+    .setAutoCollectExceptions(false)
+    .setAutoCollectPerformance(false)
+    .setAutoCollectRequests(false)
+    .setOfflineMode(true, 1)
+    .start();
+  _isStarted = true;
+  return true;
+};
+
+module.exports.trackEvent = function (eventName, properties) {
+  if (!_isStarted) return;
+  if (!properties) {
+    properties = {};
+  }
+  initPkgInfo();
+  initSubscriptionAndTenantId();
+  // Add MAC addresses & OS distro
+  if (propertyDict.mac && propertyDict.osDist && propertyDict.osRelease) {
+    internalTrackEvent(eventName, properties);
+  } else if (propertyDict.mac) {
+    initOsDistroAndTrackEvent(eventName, properties);
+  } else if (propertyDict.osDist && propertyDict.osRelease) {
+    initMacAndTrackEvent(eventName, properties);
+  } else {
+    initPropertiesAndTrackEvent(eventName, properties);

```

Comment:
How about 
if(!mac || !osDisct || !osRelease)
 initPropertiesAndTrackEvent
else
 internalTrackEnvent

That is if one info is not fetched, we fetch them all. I don't think we need to write complicated code to handle each case.